src/kaleidoscope/hooks.h

```diff
@@ -38,7 +39,34 @@ class Hooks {
   friend void ::handleKeyswitchEvent(kaleidoscope::Key mappedKey,
                                      byte row, byte col, uint8_t keyState);
 
- private:
+ public:
+
+  static EventHandlerResult onStorageReport();
+
+  template<typename T>
+  T& getFromStorage(kaleidoscope::Plugin &plugin, uint16_t offset, T& t) {
+    uint16_t plugin_offset = getStorageOffset(plugin);
+    return EEPROM.get(plugin_offset + offset, t);
+  }
+
+  template<typename T>
+  T& getFromStorage(kaleidoscope::Plugin &plugin, T& t) {
+    return getFromStorage(plugin, 0, t);
+  }
+
+  template<typename T>
+  const T& putToStorage(kaleidoscope::Plugin &plugin, uint16_t offset, T& t) {
+    uint16_t plugin_offset = getStorageOffset(plugin);
```

Comment:
I think I want a layer of indirection here. the implementation of the code should be in the hardware class. (Different microcontrollers implement persistent storage differently.)

More generally, I think we should try really hard to keep code out of the templates we use for the plugin API definitions. 

Also, do we need the multiple signature versions of each API? If we can keep the API surface down, it may stop us from boxing ourselves in later.

Reply:
> I think I want a layer of indirection here. the implementation of the code should be in the hardware class.

Agreed, these will ultimately be in the Hardware plugins. They're only here for easier review, I'll move them over once everything else is finalised, and these will be very thin wrappers.

> More generally, I think we should try really hard to keep code out of the templates we use for the plugin API definitions.

Agreed here too. By the way, in a more recent commit (which I forgot to push last night, sorry!) I moved these from `kaleidoscope:Hooks` to `Kaleidoscope.Storage.get()`/`Kaleidoscope.Storage.put()`.

> Also, do we need the multiple signature versions of each API? If we can keep the API surface down, it may stop us from boxing ourselves in later.

On second thought - no, we don't. I can just change the signature to something like:

```c++
template<typename T, typename P>
const T& put(P &plugin, T &t, uint16_t offset = 0);
```

So offset at the end with a default value, and then we don't need two variants.

Thanks!